Source/WebKit/Platform/IPC/TransferRegion.h

```diff
@@ -0,0 +1,221 @@
+#include <wtf/ThreadSafetyAnalysis.h>
+
+namespace WebCore {
+class ProcessIdentity;
+}
+
+namespace IPC {
+
+class ScopedTransferRegion;
+
+enum TransferBufferIdentifierType { };
+using TransferBufferIdentifier = ObjectIdentifier<TransferBufferIdentifierType>;
+
+class TransferRegion {
+public:
+    TransferBufferIdentifier identifier;
+    size_t size;
+    std::optional<WebKit::SharedMemory::Handle> buffer; // Null when expected to be already mapped in the target process.
+    static std::optional<TransferRegion> decode(Decoder&);
+    template<typename Encoder>
+    void encode(Encoder&) const;
+};
+
+class TransferBuffer {
+public:
+    static std::optional<TransferBuffer> create(size_t size)
+    {
+        if (!size)
+            return TransferBuffer { };
+        auto buffer = WebKit::SharedMemory::allocate(size);
+        if (!buffer)
+            return std::nullopt;
+        return TransferBuffer { buffer.releaseNonNull() };
+    }
+    TransferBuffer(TransferBuffer&&) = default;
+    TransferBuffer& operator=(TransferBuffer&&) = default;
+
+    TransferBufferIdentifier identifier() const
+    {
+        return m_identifier;
+    }
+
+    std::optional<WebKit::SharedMemory::Handle> passHandle()
+    {
+        if (m_hasBeenPassed || !m_buffer)
+            return std::nullopt;
+        m_hasBeenPassed = true;
+        return m_buffer->createHandle(WebKit::SharedMemory::Protection::ReadWrite);
+    }
+
+    std::span<uint8_t> data()
+    {
+        if (!m_buffer)
+            return { };
+        return std::span<uint8_t> { reinterpret_cast<uint8_t*>(m_buffer->data()), m_buffer->size() };
+    }
+
+    size_t size() const
+    {
+        if (!m_buffer)
+            return 0;
+        return m_buffer->size();
+    }
+
+    bool hasBeenPassed() const { return m_hasBeenPassed; }
+private:
+    TransferBuffer() = default;
+    TransferBuffer(Ref<WebKit::SharedMemory> buffer)
+        : m_buffer { WTFMove(buffer) }
+    {
+    }
+
+    TransferBufferIdentifier m_identifier { TransferBufferIdentifier::generateThreadSafe() };
+    RefPtr<WebKit::SharedMemory> m_buffer;
+    bool m_hasBeenPassed { false };
+};
+
+
+class ScopedTransferRegionMapping {
+public:
+    ScopedTransferRegionMapping(Ref<WebKit::SharedMemory> memory, size_t size)
+        : m_memory { WTFMove(memory) }
+        , m_size { size }
+    {
+    }
+    ScopedTransferRegionMapping() = default;
+
+    std::span<uint8_t> data() const
+    {
+        if (!m_memory)
+            return { };
+        return std::span<uint8_t> { reinterpret_cast<uint8_t*>(m_memory->data()), m_size };
+    }
+private:
+    RefPtr<WebKit::SharedMemory> m_memory;
+    size_t m_size { 0 };
+};
+
+// Class to hold shared memory instances that have been sent to a particular receiver process.
+class TransferRegionPool : public ThreadSafeRefCounted<TransferRegionPool> {
+    WTF_MAKE_NONCOPYABLE(TransferRegionPool);
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    explicit TransferRegionPool(Connection&);
```

Comment:
Classes that subclass RefCounted/ThreadSafeRefCounted should have a private constructor and a public create() function.